packages/smooth_app/lib/database/local_database.dart

```diff
@@ -0,0 +1,175 @@
+
+class LocalDatabase extends ChangeNotifier {
+  LocalDatabase._(final Database database) : _database = database;
+
+  final Database _database;
+
+  static const String _TABLE_PRODUCT = 'product';
+  static const String _TABLE_PRODUCT_COLUMN_BARCODE = 'barcode';
+  static const String _TABLE_PRODUCT_COLUMN_JSON = 'encoded_json';
+  static const String _TABLE_PRODUCT_COLUMN_TIMESTAMP = 'last_upsert';
+
+  static Future<LocalDatabase> getLocalDatabase() async {
+    final String databasesRootPath = await getDatabasesPath();
+    final String databasePath = join(databasesRootPath, 'smoothie.db');
+
+    final Database database = await openDatabase(
+      databasePath,
+      version: 1,
+      singleInstance: true,
+      onCreate: _onCreate,
+      onUpgrade: _onUpgrade,
+    );
+
+    return LocalDatabase._(database);
+  }
+
+  static FutureOr<void> _onCreate(final Database db, final int version) async {
+    await db.execute('create table $_TABLE_PRODUCT('
+        '$_TABLE_PRODUCT_COLUMN_BARCODE TEXT PRIMARY KEY,'
+        '$_TABLE_PRODUCT_COLUMN_JSON TEXT NOT NULL,'
+        '$_TABLE_PRODUCT_COLUMN_TIMESTAMP INT NOT NULL'
+        ')');
+  }
+
+  static FutureOr<void> _onUpgrade(
+    final Database db,
+    final int oldVersion,
+    final int newVersion,
+  ) async {
+    // not relevant for the moment, we only have 1 version of the database
+  }
+
+  static int _nowInMillis() => DateTime.now().millisecondsSinceEpoch;
+
+  Future<TableStats> getTableStats() async {
+    const String COLUMN_COUNT = 'mycount';
+    const String COLUMN_MIN = 'mymin';
+    const String COLUMN_MAX = 'mymax';
+
+    final List<Map<String, dynamic>> queryResult = await _database.query(
+      _TABLE_PRODUCT,
+      columns: <String>[
+        'count(*) as $COLUMN_COUNT',
+        'max($_TABLE_PRODUCT_COLUMN_TIMESTAMP) as $COLUMN_MAX',
+        'min($_TABLE_PRODUCT_COLUMN_TIMESTAMP) as $COLUMN_MIN',
+      ],
+    );
+    if (queryResult.isEmpty || queryResult.length > 1) {
+      // very very unlikely to happen
+      throw Exception('No stats for table $_TABLE_PRODUCT');
+    }
+    final Map<String, dynamic> uniqueRow = queryResult[0];
+    return TableStats(
+      count: uniqueRow[COLUMN_COUNT] as int,
+      minTimestamp: uniqueRow[COLUMN_MIN] as int,
+      maxTimestamp: uniqueRow[COLUMN_MAX] as int,
+    );
+  }
+
+  void dummyNotifyListeners() {
+    print('localDatabase/notifyListeners');
+    notifyListeners();
+  }
+
+  Future<Product> getProduct(final String barcode) async {
+    final List<Map<String, dynamic>> queryResult = await _database.query(
+      _TABLE_PRODUCT,
+      columns: <String>[_TABLE_PRODUCT_COLUMN_JSON],
+      where: '$_TABLE_PRODUCT_COLUMN_BARCODE = ?',
+      whereArgs: <String>[barcode],
+    );
+    if (queryResult.isEmpty) {
+      // not found
+      return null;
+    }
+    if (queryResult.length > 1) {
+      // very very unlikely to happen
+      throw Exception('Several products with the same barcode $barcode');
```

Comment:
Before the release, we should still include a possibility to choose the desired product

Reply:
The barcode is set as a Primary Key so queryResult.length will always be <= 1 :wink:

Reply:
Of course it will always be 0 or 1. Unless something very strange happens, that I prefer to catch here. It's safer and it's an explicit way to tell the future contributor: "hey, this is 0 or 1 product".

Reply:
If so, an exception is perfectly fine